src/commands/compile.rs

```diff
@@ -0,0 +1,413 @@
+                  \n\
+                  Compiling for a specific platform (Linux) and CPU preset (Skylake):\n\
+                  \n  \
+                  wasmtime compile --target x86_64-unknown-linux --skylake foo.wasm\n"
+)]
+pub struct CompileCommand {
+    #[structopt(flatten)]
+    common: CommonOptions,
+
+    /// Enable support for interrupting WebAssembly code.
+    #[structopt(long)]
+    interruptable: bool,
+
+    /// Enable SSE3 support (for x86-64 targets).
+    #[structopt(long, group = "x64")]
+    sse3: bool,
+
+    /// Enable SSSE3 support (for x86-64 targets).
+    #[structopt(long, group = "x64")]
+    ssse3: bool,
+
+    /// Enable SSE41 support (for x86-64 targets).
+    #[structopt(long, group = "x64")]
+    sse41: bool,
+
+    /// Enable SSE42 support (for x86-64 targets).
+    #[structopt(long, group = "x64")]
+    sse42: bool,
+
+    /// Enable AVX support (for x86-64 targets).
+    #[structopt(long, group = "x64")]
+    avx: bool,
+
+    /// Enable AVX2 support (for x86-64 targets).
+    #[structopt(long, group = "x64")]
+    avx2: bool,
+
+    /// Enable AVX512DQ support (for x86-64 targets).
+    #[structopt(long, group = "x64")]
+    avx512dq: bool,
+
+    /// Enable AVX512VL support (for x86-64 targets).
+    #[structopt(long, group = "x64")]
+    avx512vl: bool,
+
+    /// Enable AVX512F support (for x86-64 targets).
+    #[structopt(long, group = "x64")]
+    avx512f: bool,
+
+    /// Enable POPCNT support (for x86-64 targets).
+    #[structopt(long, group = "x64")]
+    popcnt: bool,
+
+    /// Enable BMI1 support (for x86-64 targets).
+    #[structopt(long, group = "x64")]
+    bmi1: bool,
+
+    /// Enable BMI2 support (for x86-64 targets).
+    #[structopt(long, group = "x64")]
+    bmi2: bool,
+
+    /// Enable LZCNT support (for x86-64 targets).
+    #[structopt(long, group = "x64")]
+    lzcnt: bool,
+
+    /// Enable LSE support (for aarch64 targets).
+    #[structopt(long, group = "aarch64")]
+    lse: bool,
+
+    /// Enable Nehalem preset (for x86-64 targets).
+    #[structopt(long, group = "x64", group = "preset-x64")]
+    nehalem: bool,
+
+    /// Enable Haswell preset (for x86-64 targets).
+    #[structopt(long, group = "x64", group = "preset-x64")]
+    haswell: bool,
+
+    /// Enable Broadwell preset (for x86-64 targets).
+    #[structopt(long, group = "x64", group = "preset-x64")]
+    broadwell: bool,
+
+    /// Enable Skylake preset (for x86-64 targets).
+    #[structopt(long, group = "x64", group = "preset-x64")]
+    skylake: bool,
+
+    /// Enable Cannonlake preset (for x86-64 targets).
+    #[structopt(long, group = "x64", group = "preset-x64")]
+    cannonlake: bool,
+
+    /// Enable Icelake preset (for x86-64 targets).
+    #[structopt(long, group = "x64", group = "preset-x64")]
+    icelake: bool,
+
+    /// Enable Zen preset (for x86-64 targets).
+    #[structopt(long, group = "x64", group = "preset-x64")]
+    znver1: bool,
```

Comment:
I think it'd be best if we didn't have to list out flags like this. I find it pretty unwieldy for the CLI since `wasmtime --help` becomes *super* long in not always so interesting ways.

I thought that this could be done with `--cranelift-extra-flag has_znver1=true`, but that's naturally a whole lot more wordy than `--znver1`. Could we perhaps take some liberties and make this a bit more compact though? Similar to what I'm thinking below with wasm features we could do something like:

* Backends have the ability to return all possible flags (removing the need to keep the backend in sync with this file)
* We have a general `--codegen-features=...` or something similar to Rust's `-C` flag where we do `-Ccranelift=...`
* The CLI would parse everything received and then match that against what the backend says is available, returning errors for anything enabled that the backend doesn't support
* We could implement something like `--codegen-features=help` to print out a helpful listing of all codegen features (similar to `-Ctarget-feature=help`)

I think that'd help reduce the need to keep this file in sync with the cranelift backends and also clean up the CLI UI perhaps?

Reply:
I like the idea of a `--codegen-features` option.  This will require a bit of refactoring in the generated code for Cranelift's settings as the metadata about the settings isn't suitable for describing the settings to users this way.

I'll see what I can make happen.

Reply:
Okay, so I kept the `--cranelift-set` and `--cranelift-enable` flags, removed the various Cranelift flags from `wasmtime compile`, and added a `wasmtime settings` command that prints out the available settings for the target triple, e.g.:

```
$ wasmtime settings
Cranelift settings for target 'x86_64-apple-darwin':

Boolean settings:
  has_sse3       Has support for SSE3.
  has_ssse3      Has support for SSSE3.
  has_sse41      Has support for SSE4.1.
  has_sse42      Has support for SSE4.2.
  has_avx        Has support for AVX.
  has_avx2       Has support for AVX2.
  has_avx512dq   Has support for AVX512DQ.
  has_avx512vl   Has support for AVX512VL.
  has_avx512f    Has support for AVX512F.
  has_popcnt     Has support for POPCNT.
  has_bmi1       Has support for BMI1.
  has_bmi2       Has support for BMI2.
  has_lzcnt      Has support for LZCNT.

Presets:
  baseline     A baseline preset with no extensions enabled.
  nehalem      Nehalem microarchitecture.
  haswell      Haswell microarchitecture.
  broadwell    Broadwell microarchitecture.
  skylake      Skylake microarchitecture.
  cannonlake   Canon Lake microarchitecture.
  icelake      Ice Lake microarchitecture.
  znver1       Zen (first generation) microarchitecture.

Settings enabled for this host:
  - has_sse3
  - has_ssse3
  - has_sse41
  - has_sse42
  - has_avx
  - has_avx2
  - has_popcnt
  - has_bmi1
  - has_bmi2
  - has_lzcnt
```

I've pushed up the commit, so let me know if this is inline with what you were thinking.

Reply:
The command intentionally doesn't present any of the "shared" settings as those really shouldn't be mucked around with by users, but we allow them for `--cranelift-set`/`--cranelift-enable` for users that know what they're doing.

Reply:
That looks fantastic to me, thanks for doing that!